test/src/org/exist/http/urlrewrite/URLRewritingTest.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * eXist Open Source Native XML Database
+ * Copyright (C) 2001-2018 The eXist Project
+ * http://exist-db.org
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package org.exist.http.urlrewrite;
+
+import org.apache.http.HttpResponse;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.fluent.Executor;
+import org.apache.http.client.fluent.Request;
+import org.apache.http.entity.ContentType;
+import org.exist.TestUtils;
+import org.exist.test.ExistWebServer;
+import org.exist.util.io.FastByteArrayOutputStream;
+import org.exist.xmldb.XmldbURI;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import java.io.IOException;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.exist.http.urlrewrite.XQueryURLRewrite.XQUERY_CONTROLLER_FILENAME;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class URLRewritingTest {
+
+    private static final XmldbURI TEST_COLLECTION = XmldbURI.create("/db/apps/controller-test");
+
+    private static final String TEST_CONTROLLER = "xquery version \"3.1\";\n<controller>{fn:current-dateTime()}</controller>";
+    private static Executor executor = null;
+
+    @ClassRule
+    public static final ExistWebServer existWebServer = new ExistWebServer(true, false, true, true, false);
+
+    @Test
+    public void findsParentController() throws IOException {
+        final XmldbURI nestedCollection = TEST_COLLECTION.append("nested");
+        final String docUri = getRestUri() + nestedCollection + "/test.xml";
+        final String testDocument = "<hello>world</hello>";
+
+        HttpResponse response = executor.execute(Request
+                .Put(docUri)
+                .bodyString(testDocument, ContentType.APPLICATION_XML)
+        ).returnResponse();
+        assertEquals(HttpStatus.SC_CREATED, response.getStatusLine().getStatusCode());
+
+        response = executor.execute(Request
+                .Get(docUri)
```

Comment:
URL rewriting will not be triggered if you go through the REST interface, so this line would need to request a path below `/exist/apps`:

```java
.Get(getServerUri() + "/apps/controller-test/nested/test.xml")
```